src/Grpc.AspNetCore.Server.GrpcClient/GrpcClientServiceExtensions.cs

```diff
@@ -0,0 +1,83 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Net.Http;
+using Grpc.AspNetCore.Server.GrpcClient;
+using Grpc.Core;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+using Microsoft.Extensions.Http;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    public static class GrpcClientServiceExtensions
+    {
+        public static IHttpClientBuilder AddGrpcClient<TClient>(this IServiceCollection services, Action<GrpcClientOptions<TClient>> configureClient) where TClient : ClientBase<TClient>
+        {
+            if (services == null)
+            {
+                throw new ArgumentNullException(nameof(services));
+            }
+
+            if (configureClient == null)
+            {
+                throw new ArgumentNullException(nameof(configureClient));
+            }
+
+            services.TryAdd(ServiceDescriptor.Transient(typeof(ITypedHttpClientFactory<TClient>), typeof(GrpcHttpClientFactory<TClient>)));
+            services.TryAdd(ServiceDescriptor.Transient(typeof(GrpcHttpClientFactory<TClient>.Cache), typeof(GrpcHttpClientFactory<TClient>.Cache)));
+
+            services.Configure(configureClient);
+
+            // Accessing the client options here allows the HttpContextAccessor to be added as needed
+            // Is there a way to get the value from services.Configure now?
+            var clientOptions = new GrpcClientOptions<TClient>();
+            configureClient(clientOptions);
```

Comment:
Is there a better way of doing this? I want to register the configuration with DI - `services.Configure(configureClient);` - and I want to access the configuration immediately.

Reply:
No there really isn't a way to do this, I'd recommend against running the user code here if possible. What's the core concern? 

I have a theory that this is about capturing values from the current request to use in header propagation? Using the context accessor for this is problematic because nothing enforces the constraint that you're "in a request" or "on the request processing thread.

Reply:
I wanted to access the config value here so the value could be used to decided whether AddHttpContextAccessor was added. If you want to propagate something from the current request then `AddHttpContextAccessor` would be called, otherwise no. Make it pay for play.

Reply:
Considering that the things you check for are the defaults, do you really need this?